### ISSUE TITLE
Clone enums with argless variants

### ISSUE THREAD

**Issue by amab8901** [rust-lang/rust-clippy#13120]
### What it does

Tells you to copy/clone instead of referencing when it's an enum with simple variants.

### Advantage

Enum where none of the variants have arguments takes only 1 byte to clone. But referencing takes 8 bytes. Cloning/copying is always a better idea in this case. 

### Drawbacks

Takes time to implement, and takes time for users to refactor their code in response to this lint.

### Example

```rust
#[derive(Clone, Debug)]
pub enum Hello {
    Variant1,
    Variant2,
    Variant3,
}

fn nice(hello: &Hello) {
    println!("{:?}", hello);
}

fn main() {
    let hello = Hello::Variant1;
    nice(&hello);
}

```

Could be written as:

```rust
#[derive(Clone, Debug)]
pub enum Hello {
    Variant1,
    Variant2,
    Variant3,
}

fn nice(hello: Hello) {
    println!("{:?}", hello);
}

fn main() {
    let hello = Hello::Variant1;
    nice(hello.clone());
}

```

**Comment by lukaslueg** [comment]
Same as #12786?!

**Comment by amab8901** [comment]
> Same as #12786?!

Similar but not same. This one is more concrete and specific to enums with argless variants, rather than types in general

**Comment by amab8901** [comment]
(from #12786 )
> IMHO its more desirable to enhance a developer's muscle memory with respect to not passing owned values if ownership is not required by the callee

you can also let the developer train the muscle memory of following the clippy lints

**Comment by olivia-fl** [comment]
This seems like it has some overlap with [trivially_copy_pass_by_ref](https://rust-lang.github.io/rust-clippy/master/index.html#trivially_copy_pass_by_ref). I believe believe that lint would trigger on your example code if `Hello` was `Copy`. I can't think of any situations where it makes sense to have a fieldless enum that is `Clone` but not `Copy`, so maybe a lint on missing `Copy` implementations in trivial situations would be a better approach?

**Comment by amab8901** [comment]
> This seems like it has some overlap with [trivially_copy_pass_by_ref](https://rust-lang.github.io/rust-clippy/master/index.html#trivially_copy_pass_by_ref). I believe that lint would trigger on your example code if `Hello` was `Copy`. I can't think of any situations where it makes sense to have a fieldless enum that is `Clone` but not `Copy`, so maybe a lint on missing `Copy` implementations in trivial situations would be a better approach?

great idea!

I played around with it and I found that this:

```
#![allow(clippy::let_and_return)]

#[derive()]
pub enum Hello {
    Variant1,
    Variant2,
    Variant3,
}

fn nice(hello: &Hello) -> String {
    let string = match hello {
        Hello::Variant1 => "variant1".to_string(),
        Hello::Variant2 => "variant2".to_string(),
        Hello::Variant3 => "variant3".to_string(),
    };

    string
}

fn main() {
    let hello = Hello::Variant1;
    let nice1 = nice(&hello);
    let nice2 = nice(&hello);

    dbg!(&nice1);
    dbg!(&nice2);
}

```

would indeed be good to rewrite to this:

```
#![allow(clippy::let_and_return)]

#[derive(Clone, Copy)]
pub enum Hello {
    Variant1,
    Variant2,
    Variant3,
}

fn nice(hello: Hello) -> String {
    let string = match hello {
        Hello::Variant1 => "variant1".to_string(),
        Hello::Variant2 => "variant2".to_string(),
        Hello::Variant3 => "variant3".to_string(),
    };

    string
}

fn main() {
    let hello = Hello::Variant1;
    let nice1 = nice(hello);
    let nice2 = nice(hello);

    dbg!(&nice1);
    dbg!(&nice2);
}

```

**Comment by olivia-fl** [comment]
I was imagining that the lint would trigger on the enum definition, not the use-site, so it would only trigger on enums in the same crate by construction.

**Comment by amab8901** [comment]
you mean the implementation of Clone on an enum would trigger lint to implement Copy on the same enum?

**Comment by olivia-fl** [comment]
The conservative version would be something like "lint to implement Copy on any fieldless Clone enum". A less conservative version might be "lint to implement Copy/Clone on any fieldless struct/enum", but that might have more false positives where the missing traits are intentional.

**Comment by jokeyrhyme** [comment]
Yeah, it's tricky

Say you have `pub enum A { B, C }`, and clippy tells you to make it Copy

But later you add a new variant `D(String)`

Removing the `Copy` is a backwards incompatible change from the perspective of a consumer

**Comment by olivia-fl** [comment]
That's a very good point, and applies to both of the variants I proposed.

**Comment by jokeyrhyme** [comment]
Oh, just to add: adding a new variant in an exhaustive and `pub` enum is a backwards incompatible change, so if you have to remove Copy at the same time then it's not an additional breakage

**Comment by amab8901** [comment]
I think we can start with the conservative approach and then see how it goes and whether it makes sense to extend it further

**Comment by amab8901** [comment]
> Yeah, it's tricky
> 
> Say you have `pub enum A { B, C }`, and clippy tells you to make it Copy
> 
> But later you add a new variant `D(String)`
> 
> Removing the `Copy` is a backwards incompatible change from the perspective of a consumer

perhaps we can make the lint deactivate when the enum has #[non-exhaustive] attribute? Otherwise let the lint be active by default. Adding a new variant seems comparable to adding a field to an existing variant

**Comment by jokeyrhyme** [comment]
Just to clarify my understanding: https://doc.rust-lang.org/reference/attributes/type_system.html

> The non_exhaustive attribute indicates that a type or variant may have more fields or variants added in the future.

So, I _think_ this means we are probably okay with recommending `Copy` on any fieldless `enum` with `#[derive(Clone)]` in the following cases:
- the `enum` can NOT be used outside of its crate: only the author has to deal with backwards compatibility issues
- the `enum` CAN be used outside of its crate but has `#[non_exhaustive]`: the author has already signaled to consumers to beware future backwards compatibility issues

**Comment by amab8901** [comment]
> Just to clarify my understanding: https://doc.rust-lang.org/reference/attributes/type_system.html
> 
> > The non_exhaustive attribute indicates that a type or variant may have more fields or variants added in the future.
> 
> So, I _think_ this means we are probably okay with recommending `Copy` on any fieldless `enum` with `#[derive(Clone)]` in the following cases:
> 
> * the `enum` can NOT be used outside of its crate: only the author has to deal with backwards compatibility issues
> * the `enum` CAN be used outside of its crate but has `#[non_exhaustive]`: the author has already signaled to consumers to beware future backwards compatibility issues

I agree with your suggestion. Although, I'd like to reformulate it as follows:

``` rust

fn recommend_implementing_copy_trait_on_fieldless_enum() -> bool {
    let has_nonexhaustive_attribute = has_nonexhaustive_attribute();
    let usable_outside_the_crate = usable_outside_the_crate();
    let has_clone_trait = has_clone_trait();

    let recommend_impl_copy_trait = if has_clone_trait {
        let recommend_impl_copy_trait = if has_nonexhaustive_attribute {
            true
        } else {
            !usable_outside_the_crate
        };
        recommend_impl_copy_trait
    } else {
        false
    };

    recommend_impl_copy_trait
}

```

**Comment by olivia-fl** [comment]
This seems reasonable to my understanding.
